### PR TITLE
test_runner: add --pg-version pytest argument

### PR DESCRIPTION
**Pull request by bayandin**
This PR allows to set Postgres version for testing using `--pg-version` + fixes tests for the non-default Postgres version.

Example:
```sh
./scripts/pytest -s -k test_partial_evict_tenant --pg-version 15
```

Also, it'll be easier to catch and reproduce cases like https://github.com/neondatabase/neon/issues/4032.

Test failures we have for Postgres 15 (the same we have for https://github.com/neondatabase/neon/pull/2809) + failed compatibility tests which will be in a separate PR.

#### debug build: 214 tests run: 199 passed, 2 failed, 13 ([full report](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-4037/debug/4709535149/index.html))
Failed tests:
- [`test_runner/regress/test_pg_regress.py::test_isolation`](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-4037/debug/4709535149/index.html#suites/158be07438eb5188d40b466b6acfaeb3/ec929442ac687e76) (ran [2 times](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-4037/debug/4709535149/index.html#suites/158be07438eb5188d40b466b6acfaeb3/ec929442ac687e76/retries))
- [`test_runner/regress/test_tenant_size.py::test_get_tenant_size_with_multiple_branches`](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-4037/debug/4709535149/index.html#suites/d70040d097c5c7c438e421e0416d69ce/91c9b87af94826f) (ran [2 times](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-4037/debug/4709535149/index.html#suites/d70040d097c5c7c438e421e0416d69ce/91c9b87af94826f/retries))

___
#### release build: 214 tests run: 200 passed, 1 failed, 13 ([full report](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-4037/release/4709535149/index.html))
Failed tests:
- [`test_runner/regress/test_pg_regress.py::test_isolation`](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-4037/release/4709535149/index.html#suites/158be07438eb5188d40b466b6acfaeb3/df0c62b0fae3ea2e) (ran [2 times](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-4037/release/4709535149/index.html#suites/158be07438eb5188d40b466b6acfaeb3/df0c62b0fae3ea2e/retries))

___

### CODE REVIEW

**Comment by lubennikovaav** [comment]
It's worth to mention this new flag in`test_runner/README.md`

Also, this probably need to be updated:

```
`DEFAULT_PG_VERSION`: The version of Postgres to use,
This is used to construct full path to the postgres binaries.
Format is 2-digit major version nubmer, i.e. `DEFAULT_PG_VERSION="14"`
```

File: test_runner/fixtures/pg_version.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+@enum.unique
+class PgVersion(str, enum.Enum):
+    V14 = "14"
+    V15 = "15"
+    # Instead of making version an optional parameter in methods, we can use this fake entry
+    # to explicitly rely on the default server version (could be different from pg_version fixture value)
+    NOT_SET = "<-POSTRGRES VERSION IS NOT SET->"
+
+    @staticmethod
+    def from_int(int_version: int) -> "PgVersion":
```
Comment (vadim2404) [inline test_runner/fixtures/pg_version.py:27]: what do you think about using

```python
from __future__ import annotations
```
?

I'm not insisting, just asking

Comment (bayandin) [reply test_runner/fixtures/pg_version.py:27]: Eventually, we should update Python to 3.11 and then we can use `Self` annotation.
So I'd keep it as is for now.

File: test_runner/README.md [hunk 2]
```diff
@@ -71,7 +71,8 @@ a subdirectory for each version with naming convention `v{PG_VERSION}/`.
 Inside that dir, a `bin/postgres` binary should be present.
 `DEFAULT_PG_VERSION`: The version of Postgres to use,
 This is used to construct full path to the postgres binaries.
-Format is 2-digit major version nubmer, i.e. `DEFAULT_PG_VERSION="14"`
+Format is 2-digit major version nubmer, i.e. `DEFAULT_PG_VERSION="14"`. Alternatively,
```
Comment (vadim2404) [inline test_runner/README.md:74]: In docker builds "v14" and "v15" are used. What do you think about having the same logic?

File: test_runner/fixtures/pg_version.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        return PgVersion(str(int_version)[:2])
+
+    # Make it less confusing in logs
+    def __repr__(self) -> str:
+        return f"'{self.value}'"
```
Comment (vadim2404) [inline test_runner/fixtures/pg_version.py:32]: NIT: `f"PG{self.value}"`? WDYT?

Comment (bayandin) [reply test_runner/fixtures/pg_version.py:32]: Then logs will be like this:
```
Run ['neon_local', '--pg-version', PG14, ... ]
```

This is what I meant by confusing.
So I'd stick to the current version

File: test_runner/README.md (hunk 2 above)
Comment (bayandin) [reply test_runner/README.md:74]: Ah, I always get confused where to use what 😄 
Added support for v-prefixed strings like v14/v15 as well

**Review by vadim2404 — APPROVED** [review]

**Pull request merged**